src/ray/pubsub/publisher.cc

```diff
@@ -23,53 +23,47 @@ namespace pub_internal {
 bool SubscriptionIndex::AddEntry(const std::string &key_id,
                                  const SubscriberID &subscriber_id) {
   if (key_id.empty()) {
-    // This subscriber will subscribe to all entities in the channel.
-    if (!subscribers_to_key_id_.empty()) {
-      // Skip if the channel is already in per-entity subscription mode.
-      return false;
-    }
     return subscribers_to_all_.insert(subscriber_id).second;
   }
 
-  // Skip if the channel is already in all-entity subscription mode.
-  if (!subscribers_to_all_.empty()) {
-    return false;
-  }
-
   auto &subscribing_key_ids = subscribers_to_key_id_[subscriber_id];
-  auto key_added = subscribing_key_ids.emplace(key_id).second;
+  bool key_added = subscribing_key_ids.emplace(key_id).second;
   auto &subscriber_map = key_id_to_subscribers_[key_id];
   auto subscriber_added = subscriber_map.emplace(subscriber_id).second;
 
   RAY_CHECK(key_added == subscriber_added);
   return key_added;
 }
 
-absl::optional<std::reference_wrapper<const absl::flat_hash_set<SubscriberID>>>
-SubscriptionIndex::GetSubscriberIdsByKeyId(const std::string &key_id) const {
+std::vector<SubscriberID> SubscriptionIndex::GetSubscriberIdsByKeyId(
+    const std::string &key_id) const {
+  std::vector<SubscriberID> subscribers;
```

Comment:
could it lead t some perf issue?

Reply:
I think this is unlikely to be hot, relative to other cost e.g. copying the PubMessage proto.